File: ConsoleVideoPlayer.MediaProcessor/PreProcessor.fs
module ConsoleVideoPlayer.MediaProcessor.PreProcessor

open System
open System.Diagnostics
open System.IO
open Xabe.FFmpeg

let getMetadata videoPath =
    FFmpeg.GetMediaInfo(videoPath) |> Async.AwaitTask

let extractAudio overwrite tempFolder videoPath =
    async {
        let path =
            Path.Combine(tempFolder, "ExtractedAudio")

        let pathMkv = $"%s{path}.mkv"
        let pathWav = $"%s{path}.wav"

        if overwrite then File.Delete pathWav

        let! extraction =
            FFmpeg.Conversions.FromSnippet.ExtractAudio(videoPath, pathMkv)
            |> Async.AwaitTask

        let! conversion =
            FFmpeg.Conversions.FromSnippet.Convert(pathMkv, pathWav)
            |> Async.AwaitTask

        do!
            extraction.Start()
            |> Async.AwaitTask
            |> Async.Ignore

        do!
            conversion.Start()
            |> Async.AwaitTask
            |> Async.Ignore

        File.Delete pathMkv

        return pathWav
    }

let splitIntoFrames width height overwrite tempFolder (metadata: IMediaInfo) =
    async {
        let destination = Path.Combine(tempFolder, "raw_frames")

        if overwrite then
            Directory.Delete(destination, true)

        Directory.CreateDirectory destination |> ignore

        let outputFilenameBuilder i =
            "\""
            + Path.Combine(destination, $"image%s{i}.jpg")
            + "\""

        let fnBuilderFunc =
            Func<string, string> outputFilenameBuilder

        let stream =
            (metadata.VideoStreams |> Seq.head)
                .SetCodec VideoCodec.mjpeg

        do!
            FFmpeg
                .Conversions
                .New()
                .AddStream(stream)
                .AddParameter($"-s %i{width}x%i{height}")
                .ExtractEveryNthFrame(1, fnBuilderFunc)
                .UseMultiThread(true)
                .Start()
            |> Async.AwaitTask
            |> Async.Ignore
    }

let cleanupTempDir folder =
    try
        Directory.Delete folder
    with
    | _ -> ()

let preProcess path width height tempDir =
    task {
        let sw = Stopwatch.StartNew()

        printf "Reading metadata             "
        let! metadata = getMetadata path
        printfn $"Done in %i{sw.ElapsedMilliseconds}ms"

        sw.Restart()
        printf "Preparing to pre-process     "
        cleanupTempDir tempDir
        Directory.CreateDirectory tempDir |> ignore
        printfn $"Done in %i{sw.ElapsedMilliseconds}ms"

        sw.Restart()
        printf "Extracting Audio             "
        let! audioPath = extractAudio false tempDir path
        printfn $"Done in %f{Math.Round(sw.Elapsed.TotalSeconds, 2)}s"

        sw.Restart()
        printf "Splitting into images        "
        do! splitIntoFrames width height false tempDir metadata
        printfn $"Done in %f{Math.Round(sw.Elapsed.TotalSeconds, 2)}s"

        return (metadata, audioPath)
    }

File: ConsoleVideoPlayer.MediaProcessor/PreProcessor.cs
using System.Diagnostics;
using Xabe.FFmpeg;

namespace ConsoleVideoPlayer.MediaProcessor;

public static class PreProcessor
{
	public static Task<IMediaInfo> GetMetadata(string videoPath) => FFmpeg.GetMediaInfo(videoPath);

	public static async Task<string> ExtractAudio(string videoPath, string tempDir, bool overwrite = false)
	{
		var path = Path.Combine(tempDir, "ExtractedAudio");

		var pathMkv = path + ".mkv";
		var pathWav = path + ".wav";

		if (overwrite) File.Delete(pathWav);

		await (await FFmpeg.Conversions.FromSnippet.ExtractAudio(videoPath, pathMkv)).Start();
		await (await FFmpeg.Conversions.FromSnippet.Convert(pathMkv, pathWav)).Start();

		File.Delete(pathMkv);

		return pathWav;
	}

	public static async Task SplitIntoFrames(IMediaInfo metadata, int width, int height, string tempDir,
											 bool       overwrite = false)
	{
		var dest = Path.Combine(tempDir, "RawFrames");

		if (overwrite) CleanupTempDir(dest);

		Directory.CreateDirectory(dest);

		string NameBuilder(string i) => $"\"{Path.Combine(dest, $"image{i}.png")}\"";

		var stream = metadata.VideoStreams.FirstOrDefault()?.SetCodec(VideoCodec.png);

		if (stream == null)
			throw new InvalidDataException("input media must have a video stream");

		await FFmpeg.Conversions.New()
					.AddStream(stream)
					.AddParameter($"-s {width}x{height}")
					.ExtractEveryNthFrame(1, NameBuilder)
					.UseMultiThread(true)
					.Start();
	}
	
	public static void CleanupTempDir(string path)
	{
		try { Directory.Delete(path, true); }
		catch
		{ // ignored
		}
	}

	public static async Task<(IMediaInfo, string)> PreProcess(string videoPath, string tempDir, int width, int height)
	{
		var sw = Stopwatch.StartNew();
		Console.Write("Preparing to pre-process     ");
		var metadata = await GetMetadata(videoPath);
		CleanupTempDir(tempDir);
		Directory.CreateDirectory(tempDir);
		Console.WriteLine($"Done in {sw.ElapsedMilliseconds}ms");
		
		sw.Restart();
		Console.Write("Extracting Audio             ");
		var audioPath = await ExtractAudio(videoPath, tempDir);
		Console.WriteLine($"Done in {sw.Elapsed.TotalSeconds:F2}s");
		
		sw.Restart();
		Console.Write("Splitting into images        ");
		await SplitIntoFrames(metadata, width, height, tempDir);
		Console.WriteLine($"Done in {sw.Elapsed.TotalSeconds:F2}s");

		return (metadata, audioPath);
	}
}